Mds/Collivery.php

```diff
@@ -98,850 +66,887 @@ protected function client()
+            }
+        }
+    }
+
+    /**
+     * Returns all the addresses belonging to a client.
+     *
+     * @param array $filter Filter Addresses
+     *
+     * @return array
+     * @throws Exception
+     */
+    public function getAddresses($filter = [])
+    {
+        if (($this->check_cache) && empty($filter) && $this->cache->has('collivery.addresses.'.$this->client_id)) {
+            return $this->cache->get('collivery.addresses.'.$this->client_id);
+        } else {
+            try {
+                if (empty($filter)) {
+                    $filter= ["per_page" => "0"];
+                }
+                $result = $this->consumeAPI("https://api.collivery.co.za/v3/address", $filter, 'GET');
+            } catch (Exception $e) {
+                $this->catchException($e);
+
+                return false;
+            }
+            if (isset($result['data'])) {
+                if ($this->check_cache && empty($filter)) {
+                    $this->cache->put('collivery.addresses.'.$this->client_id, $result['data'], 60 * 24);
+                }
+
+                return $result['data'];
+            } else {
+                return $this->checkError($result);
+            }
+        }
+    }
+
+    /**
+     * Returns the Contact people of a given Address ID.
+     *
+     * @param int $address_id Address ID
+     *
+     * @return array
+     * @throws Exception
+     */
+    public function getContacts($address_id)
+    {
+        if (($this->check_cache) && $this->cache->has('collivery.contacts.'.$this->client_id.'.'.$address_id)) {
+            return $this->cache->get('collivery.contacts.'.$this->client_id.'.'.$address_id);
+        } else {
+            try {
+                $result = $this->consumeAPI("https://api.collivery.co.za/v3/contacts", ["address_id" => $address_id], 'GET');
+            } catch (Exception $e) {
+                $this->catchException($e);
+
+                return false;
+            }
+
+            if (isset($result['data'])) {
+                if ($this->check_cache) {
+                    $this->cache->put('collivery.contacts.'.$this->client_id.'.'.$address_id, $result['data'], 60 * 24);
+                }
+                return $result['data'];
+            } else {
+                return $this->checkError($result);
+            }
+        }
+    }
+
+    /**
+     * Returns the POD image for a given Waybill Number.
+     *
+     * @param int $collivery_id Collivery waybill number
+     *
+     * @return array
+     * @throws Exception
+     */
+    public function getPod($collivery_id)
+    {
+        if (($this->check_cache) && $this->cache->has('collivery.pod.'.$this->client_id.'.'.$collivery_id)) {
+            return $this->cache->get('collivery.pod.'.$this->client_id.'.'.$collivery_id);
+        } else {
+            try {
+                $result = $this->consumeAPI("https://api.collivery.co.za/v3/proofs_of_delivery/", ["waybill_id" => $collivery_id, "per_page" => "0"], 'GET');
```

Comment:
Are you sure that `"per_page" => 0` is needed here? 
On clients that have several hundred delivered waybills this will become slow.
Please limit or implement pagination